### PR TITLE
UI Chart Scale Bar

### PR DESCRIPTION
**Pull request by Amstutz**
Proposal for adding UI Chart Picks. Used in the following FR: [http://www.ilias.de/docu/goto_docu_wiki_wpage_4375_1357.html](http://www.ilias.de/docu/goto_docu_wiki_wpage_4375_1357.html)

### CODE REVIEW

File: src/UI/Factory.php [hunk 1]
```diff
@@ -292,4 +292,20 @@ public function legacy($content);
 	 * @return \ILIAS\UI\Component\Panel\Factory
 	 */
 	public function panel();
+
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *     Charts are usee to graphically represent data in various forms such as maps, graphs or diagrams.
```
Comment (klees) [inline src/UI/Factory.php]: usee -> used

File: src/UI/Component/Chart/Pick.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace ILIAS\UI\Component\Chart;
+
+/**
+ * Interface Pick Charts
+ */
+interface Pick {
+
+	/**
+	 * Sets a key value pair as items for the list. Key is used as title and value is a boolean marking highlighted values.
```
Comment (klees) [inline src/UI/Component/Chart/Pick.php]: Seems not to be correct, maybe copied from somewhere?

File: src/UI/Component/Chart/Pick.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+	 * @param array string => boolean Set of elements to be rendered, boolean should be true if highlighted
+	 * @return \ILIAS\UI\Component\Listing\Descriptive
+	 */
+	public function withItems(array $items);
+
+	/**
+	 * Gets the key value pair as items for the list. Key is used as title and value as content.
```
Comment (klees) [inline src/UI/Component/Chart/Pick.php]: Seems not to be correct, maybe copied from somewhere?

File: src/UI/Component/Chart/Factory.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+/**
+ * This is how a factory for glyphs looks like.
+ */
+interface Factory {
+
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *     Pick Charts are used to display a set of items some of which especially highlighted. E.g. they can
```
Comment (klees) [inline src/UI/Component/Chart/Factory.php]: `... are used to display a selection from a set of items?` maybe?

File: src/UI/Component/Chart/Factory.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+	 *
+	 * context: >
+	 *   Pick Charts are are used in the Competence Management on the Personal Desktop.
+	 *
+	 * rules:
+	 *   composition:
+	 *     1: Each Bar of the Pick Charts MUST bear a title.
+	 *     2: The title of Pick Charts MUST NOT contain any other content than text and links.
+	 * ----
+	 * @param array string => boolean Set of elements to be rendered, boolean should be true if highlighted
```
Comment (klees) [inline src/UI/Component/Chart/Factory.php:28]: This approach (`array<string,bool>`) will break down in the moment we add another information (besides the title) to an item in the chart. Maybe we should add a more correct solution already? E.g. introducing something like a `PickChartItem`.

File: src/UI/Implementation/Component/Chart/Pick.php [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+/* Copyright (c) 2016 Timon Amstutz <timon.amstutz@ilub.unibe.ch> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\UI\Implementation\Component\Chart;
+
+use ILIAS\UI\Component as C;
+use ILIAS\UI\Implementation\Component\ComponentHelper;
+use ILIAS\UI\NotImplementedException;
+
+/**
+ * Class Descriptive
```
Comment (klees) [inline src/UI/Implementation/Component/Chart/Pick.php]: Copy Pasta.

File: src/UI/Component/Chart/Pick.php [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * Interface Pick Charts
+ */
+interface Pick {
+
+	/**
+	 * Sets a key value pair as items for the list. Key is used as title and value is a boolean marking highlighted values.
+	 * @param array string => boolean Set of elements to be rendered, boolean should be true if highlighted
+	 * @return \ILIAS\UI\Component\Listing\Descriptive
```
Comment (klees) [inline src/UI/Component/Chart/Pick.php]: Nope!

File: src/UI/Factory.php [hunk 8]
```diff
@@ -292,4 +292,20 @@ public function legacy($content);
...
+	 *   composition: >
+	 *      Charts are composed of various graphical and textual elements representing the raw data.
+	 *
+	 * rules:
+	 *   style:
+	 *      1: Charts SHOULD not rely on colors to convey information.
```
Comment (klees) [inline src/UI/Factory.php:306]: 👍

**Review by klees — APPROVED** [review]
:+1

File: src/UI/Component/Chart/Factory.php (hunk 5 above)
Comment (Amstutz) [reply src/UI/Component/Chart/Factory.php:28]: We had a similar discussion before. Key question was (and is) wheter the UI-Components should also provide data structures for their content or mostly rely on arrays. Probably we need to also provide data structures of some sort. This here might be an edge case, but e.g. trees certainly need some elaborate structure to path data into the component.

File: src/UI/Component/Chart/Pick.php (hunk 2 above)
Comment (Amstutz) [reply src/UI/Component/Chart/Pick.php]: Actually it is in my view correct. Did you comment the right line? But as pointed out before, we might change the passing data as array approach.

**Comment by atoedt** [comment]
Huhu, 
I am glad I made to this point. 

For me this git-thingy makes it very hard to understand where my scale bar went. A house wife out of her comfort zone. I will try hard... 


**If** this is a type of a family entry higher up the tree than what I called a scale bar in the FW article, **then disregard** this comment.

If this is what started out as the scale bar article I would like to ask for changes to be made: 

1) Title: Chart Pick
"Picking" is highly misleading, since in only a subset of cases users will actually pick the scale value they deem appropriate. It most cases they are presented with a scale value an appraisal or measurement, other people or objects assign to them. 

Suggestion: **Scale Bar, Scale Chart** 

2) Title: Element within bar
"Button" is wrong on several levels. We have a button in ILIAS and it is hard enough to define what a button can or cannot do. Please refrain from blurring this any further. This is not "Clicks trigger actions". 

The scale values are the elements of which a Scale Bar / Scale Chart is made of. There is nothing wrong with (I would actually perfer) to call them **Scale Values.** After all this is measureing going on. 
If you would prefer a more abstract label not relating to the purpose call it a **Pod**. 

Please consider. 

kind regards 

Alexandra

**Comment by Amstutz** [comment]
@atoedt Hi Alexandra. I appreciate your effort very much. The very close binding of our documentation to the code takes some time to get used to.

For you to get a quick overview about what is proposed, it is best to check out the factory files (see tab Files changed). For the scale bar, we propose:



> description:
>    purpose: >
>       Pick Charts are used to display a set of items some of which especially highlighted. E.g. 
>       be used to inform about a score or target on a rank ordered scale.
>    composition: >
>       Pick Charts are composed of of a set of bars of equal size. Each bar contains a title. The    highlighted elements differ from the others through their darkened background.
>  
>    context: >
>      Pick Charts are are used in the Competence Management on the Personal Desktop.
> 
>  rules:
>    composition:
>      1: Each Bar of the Pick Charts MUST bear a title.
>      2: The title of Pick Charts MUST NOT contain any other content than text and links.
> 

Note, that there is no mentioning about Buttons. It is unfortunate that I used button-like structure for the html mockup. In more recent PRs for the KS, I use only screenshots in the mockup. This simplifies things, since they are immediatly visible in the PR, e.g., see:  [(https://github.com/ILIAS-eLearning/ILIAS/pull/399/fileshttps://github.com/ILIAS-eLearning/ILIAS/pull/399/files)] for the latest item "Appointment List". So I totally agree, no buttons needed here.

Concerning the name: I agree, Scale Chart sounds better. We should discuss this in the JF.

**Comment by atoedt** [comment]
Thank you so much for the clarification. That format was much mor readable to me. 

I have gatherered the "button"-issue from

function base() {
 +	$bar1 = "<button type=\"button\" class=\"btn btn-info btn-sm\">Grade 1</button>";
 +
 +	$bar2 = "<button type=\"button\" class=\"btn btn-info btn-sm active\">Grade 2</button>";
 +
 +	$bar3 = "<button type=\"button\" class=\"btn btn-info btn-sm\">Grade 3</button>";
 +
 +	return $bar1." ".$bar2." ".$bar3;
 +}

**Comment by Amstutz** [comment]
Renamed Pick Chart to Scale Bar

**Comment by matthiaskunkel** [comment]
Jour Fixe, 16 Jan 2017: We highly appreciate this pull request. Please commit it to the trunk. The scale elements should be ordered horizontally for mobile use, too.

**Comment by chfsx** [comment]
@Amstutz this PR currently has conflicts

**Comment by Amstutz** [comment]
@atoedt ,@alex40724 : Will this be implemented for 5.3? If not, are there plans for this to be implemented at all? If not, we should close this.

**Comment by alex40724** [comment]
Yes, this will be implemented.

**Comment by alex40724** [comment]
Closing this one. Follow-Up with implementation is #572.

**Pull request closed without merging**